gtk/Session.cc

```diff
@@ -905,18 +907,62 @@ void Session::torrent_changed(tr_torrent_id_t id)
 
 void Session::remove_torrent(tr_torrent_id_t id, bool delete_files)
 {
+    struct CallbackUserData
+    {
+        std::weak_ptr<Impl> impl;
+        tr_torrent_id_t id;
+        bool succeeded = false;
+    };
+
     if (auto const& [torrent, position] = impl_->find_torrent_by_id(id); torrent)
     {
-        /* remove from the gui */
-        impl_->get_raw_model()->remove(position);
+        // This is the callback called from the Gtk main thread. If successfull,
+        // it removes the torrent entry from the GUI.
+        auto callback_this_thread = [](gpointer user_data) -> gboolean
+        {
+            auto ud = std::unique_ptr<CallbackUserData>(static_cast<CallbackUserData*>(user_data));
+            if (ud->succeeded)
+            {
+                if (auto const impl = ud->impl.lock(); impl)
+                {
+                    if (auto const& [torrent_, position_] = impl->find_torrent_by_id(ud->id); torrent_)
+                    {
+                        /* remove from the gui */
+                        impl->get_raw_model()->remove(position_);
```

Comment:
As I pointed out earlier (https://github.com/transmission/transmission/pull/6055#issuecomment-1877592011), making torrent inaccessible during the removal might be a good idea. Not sure what will happen if user tries to perform some other operations on a torrent during this time, including trying to remove it again, but if nothing critical (e.g. such operations will be failing / libtr will be returning errors) then I'd be okay with how it is for now, and we'll work on making UI reflect the state separately.

Reply:
At least the Qt and daemon clients don't do anything to the entry while it is being removed, so I don't think this is a problem here.